File: v2/rp_runtime/src/lib/narrative-visibility-parse.mjs
/** Parse combined narrator/opening JSON envelopes with narrative_visibility (#81). */

export function parseNarratorVisibilityEnvelope(raw) {
  const text = String(raw ?? '').trim();
  if (!text) {
    return { presentationText: null, narrativeVisibility: null, parseError: 'empty output' };
  }

  let stripped = text;
  if (stripped.startsWith('```')) {
    stripped = stripped.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/, '').trim();
  }

  if (!stripped.startsWith('{')) {
    return { presentationText: stripped, narrativeVisibility: null, parseError: null };
  }

  try {
    const payload = JSON.parse(stripped);
    if (!payload || typeof payload !== 'object') {
      return { presentationText: stripped, narrativeVisibility: null, parseError: 'envelope not an object' };
    }
    const presentationText = String(
      payload.presentation_text ?? payload.presentation ?? payload.text ?? '',
    ).trim() || null;
    const narrativeVisibility = payload.narrative_visibility;
    if (narrativeVisibility != null && typeof narrativeVisibility !== 'object') {
      return { presentationText, narrativeVisibility: null, parseError: 'narrative_visibility not an object' };
    }
    return { presentationText, narrativeVisibility: narrativeVisibility ?? null, parseError: null };
  } catch {
    return { presentationText: stripped, narrativeVisibility: null, parseError: null };
  }
}

File: v2/domain/modules/narrative_visibility_parse.py
"""Parse combined narrator/opening JSON envelopes with narrative_visibility (#81)."""

from __future__ import annotations

import json
import re
from typing import Any


def _strip_code_fence(raw: str) -> str:
    text = raw.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text, count=1)
        text = re.sub(r"\s*```$", "", text)
    return text.strip()


def parse_narrator_visibility_envelope(raw: str) -> tuple[str | None, dict[str, Any] | None, str | None]:
    """Return (presentation_text, narrative_visibility_dict, error)."""
    if not raw or not str(raw).strip():
        return None, None, "empty output"

    stripped = _strip_code_fence(str(raw))
    if not stripped.startswith("{"):
        return stripped, None, None

    try:
        payload = json.loads(stripped)
    except json.JSONDecodeError:
        return stripped, None, None

    if not isinstance(payload, dict):
        return stripped, None, "envelope not an object"

    presentation = payload.get("presentation_text")
    if presentation is None:
        presentation = payload.get("presentation")
    if presentation is None and isinstance(payload.get("text"), str):
        presentation = payload.get("text")

    presentation_text = str(presentation or "").strip() or None
    nvr = payload.get("narrative_visibility")
    if nvr is not None and not isinstance(nvr, dict):
        return presentation_text, None, "narrative_visibility not an object"

    units = nvr.get("units") if isinstance(nvr, dict) else None
    if units is not None and not isinstance(units, list):
        return presentation_text, None, "narrative_visibility.units not a list"

    return presentation_text, (nvr if isinstance(nvr, dict) else None), None
